### PR TITLE
Convert requirementslib to pydantic

### PR DESCRIPTION
**Pull request by matteius**
Converts requirementslib to pydantic away from attr/attrs.
* Stop importing setuptools
* Remove unused code
* Require lines being parsed to be valid requirements, because we cannot reasonably infer versions and proper names from all lines -- think of VCS requirements, the repository name may not match the setup.py name, or the egg name could differ.

We need to continue overhauling requirementslib after these changes, there are some long standing known issues that this branch does not address:
* detect extras when installing whl files
* use the updated pip line formatting (egg fragments are deprecated)

### CODE REVIEW

File: src/requirementslib/fileutils.py [hunk 1]
```diff
@@ -92,19 +92,9 @@ def get_long_path(short_path: Text) -> Text:
 
 
 def normalize_path(path):
-    """Return a case-normalized absolute variable-expanded path.
```
Comment (oz123) [inline src/requirementslib/fileutils.py:95]: I'd like us to keep doc strings in place, if you are not adding type annotations.

File: src/requirementslib/models/metadata.py [hunk 2]
```diff
@@ -382,38 +338,41 @@ def from_str(cls, depstr, parent=None):
...
+    def check_value(self, algorithm, value):
+        self.check_algorithm(algorithm)
+        validate_digest(algorithm, value)
+        return value
+
 
 # XXX: This is necessary because attrs converters can only be functions, not classmethods
-def create_digest_collection(digest_dict):
-    # type: (TDigestDict) -> List["Digest"]
+def create_digest_collection(digest_dict) -> List["Digest"]:
```
Comment (oz123) [inline src/requirementslib/models/metadata.py:375]: The comment above this line is no longer relevant.

**Review by oz123 — CHANGES_REQUESTED** [review]
Your first commit removed cache.py and resolvers.py but it does not tell much.
Can you explain a bit why these were removed?

Did you migrate the functionality to other modules?

**Comment by matteius** [comment]
@oz123 

> Your first commit removed cache.py and resolvers.py but it does not tell much.
Can you explain a bit why these were removed?
Did you migrate the functionality to other modules?

It was the prep work I did explained here:  https://github.com/pypa/pipenv/pull/5663

File: src/requirementslib/fileutils.py (hunk 1 above)
Comment (matteius) [reply src/requirementslib/fileutils.py:95]: but the implementation changed and I wasn't sure if the comment is still accurate.
Willing to add it back though for me it didn't provide much value.

Comment (oz123) [reply src/requirementslib/fileutils.py:95]: ok.

**Pull request merged**